File: src/services/dap/launch.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";

import { debuggeeJavaArgs, jdwpAgentArg } from "./launch.ts";

test("debuggeeJavaArgs orders agent, vm args, classpath, main class, program args", () => {
  const args = debuggeeJavaArgs("/cp", "example.App", {
    vmArgs: ["-Xmx64m", "-Dk=v"],
    programArgs: ["a", "b"],
  });
  assert.deepEqual(args, [
    jdwpAgentArg(),
    "-Xmx64m",
    "-Dk=v",
    "-cp",
    "/cp",
    "example.App",
    "a",
    "b",
  ]);
});

test("debuggeeJavaArgs omits empty vm/program args but keeps -cp and main class", () => {
  assert.deepEqual(debuggeeJavaArgs("/cp", "M"), [jdwpAgentArg(), "-cp", "/cp", "M"]);
});

test("vm args precede the main class so the JVM (not the program) receives them", () => {
  const args = debuggeeJavaArgs("/cp", "M", { vmArgs: ["-ea"], programArgs: ["-ea"] });
  // Both literally "-ea": the first (JVM flag) is before "M", the program one after.
  assert.ok(args.indexOf("-ea") < args.indexOf("M"));
  assert.ok(args.lastIndexOf("-ea") > args.indexOf("M"));
});

File: togo/internal/dapserver/launch_test.go
package dapserver

import (
	"slices"
	"testing"
)

func TestDebuggeeJavaArgsOrdering(t *testing.T) {
	got := DebuggeeJavaArgs("/cp", "example.App", LaunchOptions{
		VMArgs:      []string{"-Xmx64m", "-Dk=v"},
		ProgramArgs: []string{"a", "b"},
	})
	want := []string{jdwpAgentArg, "-Xmx64m", "-Dk=v", "-cp", "/cp", "example.App", "a", "b"}
	if !slices.Equal(got, want) {
		t.Fatalf("got %v want %v", got, want)
	}
}

func TestDebuggeeJavaArgsMinimal(t *testing.T) {
	got := DebuggeeJavaArgs("/cp", "M", LaunchOptions{})
	want := []string{jdwpAgentArg, "-cp", "/cp", "M"}
	if !slices.Equal(got, want) {
		t.Fatalf("got %v want %v", got, want)
	}
}

func TestDebuggeeJavaArgsVMArgsBeforeMainClass(t *testing.T) {
	got := DebuggeeJavaArgs("/cp", "M", LaunchOptions{VMArgs: []string{"-ea"}, ProgramArgs: []string{"-ea"}})
	// Both literally "-ea": the JVM flag precedes the main class, the program
	// arg (the last element) follows it.
	if slices.Index(got, "-ea") >= slices.Index(got, "M") {
		t.Fatalf("vm arg not before main class: %v", got)
	}
	if got[len(got)-1] != "-ea" {
		t.Fatalf("program arg not after main class: %v", got)
	}
}
